pkg/catalog/manager/manager_test.go

```diff
@@ -53,17 +53,172 @@ func TestLatestAvailableTemplateVersion(t *testing.T) {
 	testLatestAvailableTemplateVersion(t, "dev", "0.12.16", template)
 	testLatestAvailableTemplateVersion(t, "master", "0.12.16", template)
 	testLatestAvailableTemplateVersion(t, "master-head", "0.12.16", template)
 	testLatestAvailableTemplateVersion(t, "", "0.12.16", template)
 
+	testLatestAvailableTemplateVersion(t, "v2.1.0-alpha1", "0.12.16", templateWithoutRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.0.1", "0.12.16", templateWithoutRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.0.2-beta", "0.12.16", templateWithoutRancherVersion)
 	testLatestAvailableTemplateVersion(t, "v2.1.0", "0.12.16", templateWithoutRancherVersion)
-	testLatestAvailableTemplateVersion(t, "dev", "0.12.16", templateWithoutRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.2.5", "0.12.16", templateWithoutRancherVersion)
 	testLatestAvailableTemplateVersion(t, "master", "0.12.16", templateWithoutRancherVersion)
 	testLatestAvailableTemplateVersion(t, "master-head", "0.12.16", templateWithoutRancherVersion)
 	testLatestAvailableTemplateVersion(t, "", "0.12.16", templateWithoutRancherVersion)
+
+	testLatestAvailableTemplateVersion(t, "v2.1.0-alpha1", "0.12.16", templateWithoutMinRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.0.1", "0.12.16", templateWithoutMinRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.0.2-beta", "0.12.16", templateWithoutMinRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.1.0", "0.12.16", templateWithoutMinRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.2.5", "0.12.16", templateWithoutMinRancherVersion)
+	testLatestAvailableTemplateVersion(t, "dev", "0.12.16", templateWithoutMinRancherVersion)
+	testLatestAvailableTemplateVersion(t, "master", "0.12.16", templateWithoutMinRancherVersion)
+	testLatestAvailableTemplateVersion(t, "master-head", "0.12.16", templateWithoutMinRancherVersion)
+	testLatestAvailableTemplateVersion(t, "", "0.12.16", templateWithoutMinRancherVersion)
+
+	testLatestAvailableTemplateVersion(t, "v2.1.0-alpha1", "0.12.14", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.0.1", "0.12.14", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.0.2-beta", "0.12.14", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.1.0", "0.12.15", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.2.5", "0.12.16", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "dev", "0.12.16", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "master", "0.12.16", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "master-head", "0.12.16", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "", "0.12.16", templateWithoutMaxRancherVersion)
+
+	testLatestAvailableTemplateVersion(t, "v2.2.0-alpha1", "0.12.16", templateWithMinPrerelease)
```

Comment:
There is no test case here that expects 0.12.15. However, it doesn't seem possible as the range for 0.12.16 (`[v2.2.0-alpha1, v2.3.0]`) fully includes the range for 0.12.15 (`[v2.2.0-alpha1, v2.2.0]`) so 0.12.15 won't ever be used.

Now, these ranges don't really match what we're generally using per the examples you provided in your comment https://github.com/rancher/rancher/pull/35507/files#r751808450.

`templateWith99MaxPatchVersion` does match one of these examples, but neither `templateWithMinPrerelease` nor `templateWithMinAndMaxPrerelease` match "x.y.z-rc1 to x.y.(z+1)" example. So I would recommend replacing these 2 sets with new set that has ranges like `[2.2.0-rc1, 2.2.1]` and `[2.2.1-rc1, 2.2.2]` and have a few test cases around these versions.

Reply:
Done, also added more tests to the 99 version to highlight the case for ranges like `>=v2.2.0-alpha1,<=v2.2.99` being followed up by `>=v2.3.0-alpha1,<=v2.3.99` (`v2.3.0-0` should not be accepted by either).